### PR TITLE
WebAuthn extension

### LINKED ISSUE #9316: Web Authn Extension
**Describe the extension**
Have been planning to open the issue for a while:
https://en.wikipedia.org/wiki/WebAuthn

I've noticed recently Vertx Auth ships the extension (cc @pmlopes) so we can have an extension which builds on top of it. And then may be recommend it as the main option when no SSO is involved

Interested in this extension, please +1 via the emoji/reaction feature of GitHub (top right).

CC @stuartwdouglas @pedroigor

### PR DESCRIPTION
**Pull request by FroMage**
This is the infamous Quarkus WebAuthn extension, finally.

I'm making this a draft PR because since my rebase on the latest Quarkus, it fails due to a call to `Context.get` in the Vert.x WebAuthn extension, which hopefully @cescoffier and @pmlopes are going to help me with.

Also, I'll need @pmlopes's help with the documentation for the `WebAuthnOptions` which I'm exposing as Quarkus config: they're undocumented so I can't document them.

Once this is resolved I can turn this out of draft, but otherwise the tests are there, passing (well, minus the context issue), and the docs are written, and the quickstart PR is on its way.

Also, I have a question for @stuartwdouglas and @geoand: here I'm shipping a copy of `RenardeCookieFilter` which is a fix/fork of the RestAssured `CookieFilter` which has an invalid cookie expiry date parser that actually doesn't work. I'd like to add this class to an existing Quarkus test module so that everyone can use it, but I'm not sure which would be suitable, do you know?

Fixes #9316

### CODE REVIEW

**Comment by pmlopes** [comment]
@FroMage I don't have write permissions on quarkus, so I'll make a fork from your repo and do PR's against your branch so you can cherry pick and push to quarkus. Is this OK?

**Comment by FroMage** [comment]
Sure, thanks :)

**Comment by cescoffier** [comment]
@pmlopes One of the problems @FroMage is seeing is that `VertxContextPRNG` is using `Context.put`, which is forbidden in Quarkus. We had many issues with bad usage of `Context.get` / `Context.put` methods, leaking sensible data between unrelated processing. 

```
2022-03-15 11:06:45,608 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (vert.x-eventloop-thread-0) HTTP Request to /webauthn/register failed, error id: 38925640-db93-4129-aa7a-05207437e5e2-1: java.lang.UnsupportedOperationException: Access to Context.put(), Context.get() and Context.remove() are forbidden as it can leak data between unrelated processing. Use Context.putLocal(), Context.getLocal() and Context.removeLocal() instead. Note that these methods can only be used from a 'duplicated' Context, and so may not be available everywhere.
    at io.quarkus.vertx.core.runtime.VertxLocalsHelper.throwOnRootContextAccess(VertxLocalsHelper.java:18)
    at io.vertx.core.impl.AbstractContext.get(AbstractContext.java)
    at io.vertx.ext.auth.VertxContextPRNG.current(VertxContextPRNG.java:66)
    at io.vertx.ext.auth.VertxContextPRNG.current(VertxContextPRNG.java:98)
    at io.vertx.ext.auth.webauthn.impl.WebAuthnImpl.<init>(WebAuthnImpl.java:65)
    at io.vertx.ext.auth.webauthn.WebAuthn.create(WebAuthn.java:55)
    at io.quarkus.security.webauthn.WebAuthnSecurity.<init>(WebAuthnSecurity.java:71)
    at io.quarkus.security.webauthn.WebAuthnSecurity_Bean.create(Unknown Source)
    at io.quarkus.security.webauthn.WebAuthnSecurity_Bean.create(Unknown Source)
```

**Comment by geoand** [comment]
> I'd like to add this class to an existing Quarkus test module so that everyone can use it, but I'm not sure which would be suitable, do you know?

I don't think we have a test module for HTTP related stuff

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRunTimeConfig.java
Comment (pmlopes) [inline extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRunTimeConfig.java:31]: This is double configuration and should be removed.

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRunTimeConfig.java
Comment (pmlopes) [inline extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRunTimeConfig.java:89]: Don't enable this until: https://github.com/vert-x3/vertx-auth/issues/535 is fixed as it implies a behavior change upstream.

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRunTimeConfig.java
Comment (pmlopes) [inline extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRunTimeConfig.java:92]: You could enable this like in vert.x, you can use the "standard" base64 certificate encoder string:

https://github.com/vert-x3/vertx-auth/blob/master/vertx-auth-webauthn/src/main/java/io/vertx/ext/auth/webauthn/WebAuthnOptions.java#L47-L184

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRunTimeConfig.java
Comment (pmlopes) [inline extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRunTimeConfig.java:95]: This one is like the previous. Without this, MDS doesn't work. MDS is important for PSD2.0 apps (banking/finance) where authenticators must be verified, for example, a bank should not trust a rooted android, as a bad actor could attempt to extract the private key from the hardware in case of security bugs.

File: extensions/security-webauthn/runtime/src/main/resources/webauthn.js
Comment (pmlopes) [inline extensions/security-webauthn/runtime/src/main/resources/webauthn.js:1]: Perhaps you should just include this from maven?

https://repo1.maven.org/maven2/io/vertx/vertx-auth-webauthn/4.2.5/vertx-auth-webauthn-4.2.5-client.js

This will be updated once the new Level 3 spec is released as we hope that the API will handle JSON directly instead of this dance of encode/decode buffers.

Plus we need to support https://github.com/vert-x3/vertx-auth/issues/539 and you will start drifting from the original script and have double maintenance

**Review by pmlopes — CHANGES_REQUESTED** [review]
I've added some comments at the high-level config and JS script. Please have a look at the downstream PR that adds documentation.

**Comment by pmlopes** [comment]
PRNG should be thread safe as it relies on `SecureRandom` which is thread safe. Feel free to review:

https://github.com/vert-x3/vertx-auth/blob/master/vertx-auth-common/src/main/java/io/vertx/ext/auth/PRNG.java

The only extra thing PRNG is doing, is that it sets up a timer that reseeds the secure random, as described by the OWASP security best practices.

File: extensions/security-webauthn/runtime/src/main/resources/webauthn.js
Comment (FroMage) [reply extensions/security-webauthn/runtime/src/main/resources/webauthn.js:1]: I have updated it with more methods, to be able to only do the first half and get the login/registration challenges without calling the callback.

**Comment by pmlopes** [comment]
@FroMage @cescoffier please review https://github.com/vert-x3/vertx-auth/pull/546 it should address the context issue and work not only for webauthn but for all vert.x code that uses it.

File: docs/src/main/asciidoc/security-webauthn.adoc
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-webauthn.adoc:1300]: @FroMage This is a very good guide - can you please add a micro section to `security.adoc` itself linking to this guide, above the `OIDC` section there, since `security.adoc` is like a dashboard pointing to all the security related docs.

The other thing which I believe is important to do now is to add a short note/section to this new guide clarifying that it does not replace OIDC but rather, as you mentioned the other day, introduces a password less authentication which can be used with or without OIDC as a way to address an authentication challenge.

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnAuthenticationMechanism.java
Comment (sberyozkin) [inline extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnAuthenticationMechanism.java:75]: A non null type has to be returned here, otherwise it won't work alongside other mechanisms

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnAuthenticationMechanism.java
Comment (sberyozkin) [inline extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnAuthenticationMechanism.java:46]: I wonder if `WebAuthn` specific request type should be introduced, as `Trusted` one might be only applicable to the MTLS mechanism, @stuartwdouglas, what do you think ?

Comment (sberyozkin) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnAuthenticationMechanism.java:46]: Actually, it is already available, so I think it should be used instead, and this mechanism should not advertize it can support a `Trusted` one

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnController.java
Comment (sberyozkin) [inline extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnController.java:229]: This is code is executed outside of `WebAuthn` `HttpAuthenticationMechanism` - which I think should be the entry point for `WebAuthn` authenticating the user

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnAuthenticationMechanism.java
Comment (FroMage) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnAuthenticationMechanism.java:46]: AFAIK this is actually only used for the cookie session stored after Form-based auth. Not MTLS.

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnAuthenticationMechanism.java
Comment (FroMage) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnAuthenticationMechanism.java:75]: What sort of value should it return, then? I'm not sure how this is applicable at all.

File: docs/src/main/asciidoc/security-webauthn.adoc
Comment (FroMage) [reply docs/src/main/asciidoc/security-webauthn.adoc:1300]: Good point.

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnController.java
Comment (FroMage) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnController.java:229]: You mean the callbacks are defined as autonomous routes and not as part of the `HttpAuthenticationMechanism` interception like OIDC? I thought that was really weird and confusing to do, when they have their own URI mapping and can't really interfere in other requests, unlike cookie or header-based auth like Basic Auth session cookies.
Is there a real benefit in doing it the complex way? This seems much cleaner.

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRunTimeConfig.java
Comment (FroMage) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRunTimeConfig.java:170]: I'll leave this open until someone wants this. I don't feel good telling people to add base64 certificates in the config file, that's just bad UX.
If anything, I'd prefer the config to point at files:

```properties
quarkus.webauthn.root-certificates.something=something.pem
```

Or even a convention where these files get automatically configured `src/main/resources/META-INF/webauthn/root/*.pem`

Comment (FroMage) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRunTimeConfig.java:170]: BTW, what's the point of the `Map`? Why not just a list?

**Comment by FroMage** [comment]
… Which… has been released !

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRunTimeConfig.java
Comment (pmlopes) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRunTimeConfig.java:163]: The reason for the map was because older versions of vertx didn't supported all attestation modes and the root certificates are to be bound to a given attestation type. Internally vert.x uses a service loader to load the attestation implementations, so this was dynamic.

Perhaps we can improve the upstream code to not use a service loader as it now supports all attestations so we can define speicifc config properties, e.g.:

* android-safetynet
* android-key
* apple

And `mds` (metadata service).

We could have a dual option too: either inline buffer (which you don't like) and a filesystem path.

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRunTimeConfig.java
Comment (pmlopes) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRunTimeConfig.java:169]: Same as before, we can propose an upstream update that takes a list of files.

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRunTimeConfig.java
Comment (FroMage) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRunTimeConfig.java:163]: We can handle files downstream, that's no problem. But I still don't understand why this is a `Map`, do you do anything special with the keys?

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java
Comment (FroMage) [inline extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java:41]: I'm guessing these paths should start with `/q` perhaps?

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRunTimeConfig.java
Comment (pmlopes) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRunTimeConfig.java:163]: yes, they need to match the attestation types:

* android-safetynet
* android-key
* apple

This way during the attestation we know which certificate to use.

**Comment by FroMage** [comment]
Awesome! That'd be really great!

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnAuthenticationMechanism.java
Comment (sberyozkin) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnAuthenticationMechanism.java:75]: It can be anything like new `HttpCredentialTransport(HttpCredentialTransport.Type.WEBAUTHN, "webauthn");` for a start,
then people can have a path specific authentication, for example, `webauthn` but for some paths `bearer`. Not urgent I guess but it can be worth it

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnAuthenticationMechanism.java
Comment (sberyozkin) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnAuthenticationMechanism.java:46]: OK, these are mostly the internal details for now so I guess we can change things if needed later

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnController.java
Comment (sberyozkin) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnController.java:233]: @FroMage What I'm confused about (or concerned is probably the better word) is that we have several paths for obtaining the identity, can you clarify please when is the `WebAuthn` auth mechanism is used and when via these autonomous routes ?
Perhaps it is a neater way to do it, but I'd say we should review how to do it consistently across all of the mechanisms.

**Comment by sberyozkin** [comment]
Hi @FroMage Sorry for a delay, IMHO it is quite close but I also think we should have 2 approvals here, @stuartwdouglas hi Stuart, can you also have a look please ?

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnAuthenticationMechanism.java
Comment (FroMage) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnAuthenticationMechanism.java:75]: Yeah, not urgent, because not obvious. I could use `POST` because the callback that ships using `POST` with JSON, but users can also provide their own callbacks and I've used `POST` with HTTP forms. Really not sure what'd be applicable here. Unless I add a type as you recommend, but I'm not sure.

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnAuthenticationMechanism.java
Comment (FroMage) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnAuthenticationMechanism.java:46]: Cool

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnController.java
Comment (sberyozkin) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnController.java:233]: @FroMage I think I understand it better now after looking through the code again. But the challenge is different here, with `WebAuthnAuthenticationMechanism` the failure results in a redirect to the login page challenge - but with the custom callback it is 401 - is it expected for this use case ?

**Comment by sberyozkin** [comment]
@gsmet Thanks, it feels like we need another day or so

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java
Comment (sberyozkin) [inline extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java:26]: @FroMage Is this key intended to be used in dev mode only ? If yes, then I think the right way to handle it to have it generated in the `deployment` as the `DevServices for WebAuthn` and then set as a runtime configuration property

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java
Comment (sberyozkin) [inline extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java:55]: I'm not sure it is a good idea to have that reused, I mean, it is unlikely people would prefer to use `FormAuth` (which is what that key configuration was intended for) alongside `WebAuthn`  - but nonetheless I think it will be cleaner to have a similar property in `WebAuthnRunTimeConfig`

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java
Comment (sberyozkin) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java:41]: Steph, can you explain a bit more about these routes - are they all protected, or can a user without a WebAuthn registration invoke on any of these routes ?

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/PersistentLoginManager.java
Comment (sberyozkin) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/PersistentLoginManager.java:107]: @stuartwdouglas Stuart, can you please review these `PersistenLoginManager` changes ?

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnController.java
Comment (FroMage) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnController.java:233]: I don't think so: with a custom endpoint I use `WebAuthnSecurity.login/register` and I get a `Uni` with an exception and then it's up to the custom endpoint to figure out what to do.

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java
Comment (FroMage) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java:26]: I don't think so, this is a direct copy of how it's done for Basic FORM auth.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/PersistentLoginManager.java
Comment (FroMage) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/PersistentLoginManager.java:107]: Not sure this is a big change, I'm just allowing this to be reused for other cookie names.

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java
Comment (geoand) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java:41]: If they are framework routes, they should probably under `/q`

Comment (FroMage) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java:41]: Yeah, I thought the same.

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnAuthenticationMechanism.java
Comment (sberyozkin) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnAuthenticationMechanism.java:75]: @FroMage `HttpCredentialTransport` was initially added to avoid several mechanisms with the same type being registered; the first value can be anything really, but it is a string like `webauthn` which will be useful as the users do use path specific authentications, so it would help with https://quarkus.io/guides/security#path-specific-authentication-mechanism. A non null value will need to be added later on to support it but it can indeed be done later; I might add a test later

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java
Comment (sberyozkin) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java:41]: This path, `q` or something else, is configurable, I don't recall the name, something with the `non-application` qualifier

Comment (geoand) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java:41]: See `NonApplicationRootPathBuildItem`

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnController.java
Comment (sberyozkin) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnController.java:233]: @FroMage It does look like it terminates with 401 here, https://github.com/quarkusio/quarkus/pull/24316/files#diff-171ecf0519a39e39e4ae595be50469f3e93f0caf49494561f0075e7b05330ddeR254 ?

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java
Comment (sberyozkin) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java:41]: @FroMage 

> These routes are not protected: they're the only way in to authenticate and register via webauthn.
But did you explain earlier `WebAuthnAuthenticationMechanism` was an alternative to some of these autonomous routes ? 

and re the `/register` route - (I read the docs but likely missed something) where is that public key saved ? The docs says it is not saved anywhere, so I just wonder how is the signature verified.

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java
Comment (sberyozkin) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java:55]: @FroMage but `WebAuthn` and `Form` are two different mechanisms which can co-exist, so even though both need the encryption key, it is generally not a good idea to use the same key material for different targets/purposes. The only thing, as I said, it appears very unlikely these 2 mechanisms can co-exist in a production but Quarkus allows for it. I'd really like @stuartwdouglas to comment on re-using the config/code intended for Form originally for WebAuthn. But it is a `preview` so I guess it is not a blocker for a merge

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/PersistentLoginManager.java
Comment (sberyozkin) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/PersistentLoginManager.java:107]: It is fine, my concern is about a case where both WebAuthn and Form mechanisms can coexist; but lets address it before considering a status change from `preview`

**Comment by sberyozkin** [comment]
@FroMage Can you please start squashing the commits ?

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnController.java
Comment (sberyozkin) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnController.java:233]: Never mind - that is a bad request 400

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java
Comment (FroMage) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java:55]: From my PoV, every encrypted/signed cookie from Quarkus should use the same algo/key/settings. Because I really can't imagine why we'd want different settings for different cookies.

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRunTimeConfig.java
Comment (pmlopes) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRunTimeConfig.java:163]: For reference, here's the service loader config:

https://github.com/vert-x3/vertx-auth/blob/master/vertx-auth-webauthn/src/main/resources/META-INF/services/io.vertx.ext.auth.webauthn.impl.attestation.Attestation

And the names are coming from the interface:

https://github.com/vert-x3/vertx-auth/blob/master/vertx-auth-webauthn/src/main/java/io/vertx/ext/auth/webauthn/impl/attestation/Attestation.java#L37-L41

File: extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java
Comment (FroMage) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java:41]: Done for the `/q` prefix

Comment (sberyozkin) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java:41]: @FroMage I suppose I was confused by `So the association of the hardware authenticator and the target service will always derive the same private and public key pair, none of which are stored anywhere`, but indeed the docs describe it, sorry

Comment (sberyozkin) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java:41]: @FroMage 

> WebAuthnAuthenticationMechanism is just an implementation detail.

I don't think it is a correct statement - it implements `HttpAuthenticationMechanism` - which means it should drive the authentication on its own without the user having to start from `/webauthn/login`. I.e, if we have 
```
@Path("myservice")
class MyEndpoint { @GET String get() {return "hello";} }
```
then `WebAuthnAuthenticationMechanism` should drive the authentication when a user accesses `http://localhost:8080/myservice`. I'm looking at the code and it appears to be the case - i.e, it challenges the user with a redirect to `/login`

Comment (sberyozkin) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java:41]: @FroMage, Steph, I must be slow, sorry, so given this flow I've just described, a few questions:
- do you expect it to work the way I've described
- what happens to the original `http://localhost:8080/myservice` request - should the user be expected to be returned to `http://localhost:8080/myservice` ? (I don't see it is a redirect back in the `/login` handler, not a bit problem, just curious)

Comment (FroMage) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java:41]: Yes, it should redirect you to the login page, which the user has to implement.

And no, it doesn't redirect you back to the original page ATM, thought I guess we could add support for that in the future, if someone asks for it.

Comment (sberyozkin) [reply extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnRecorder.java:41]: @FroMage OK, thanks. I think `WebAuthn` is a better `Form Auth` - it basically has the same kind of flow, with the main thing which is different is no password is sent.

**Comment by sberyozkin** [comment]
@gsmet Can you please review the list of commits and agree with @FroMage which of them should not be squashed, thanks

**Review by sberyozkin — APPROVED** [review]
Thanks @FroMage and @pmlopes for working on this exciting new feature; as we've discussed with Steph, some further cleanup/minor improvements may be necessary before its `preview` status gets upgraded, but indeed, Quarkus users should get a chance to start experimenting with this feature asap

**Comment by sberyozkin** [comment]
@FroMage I think we should indeed position `WebAuthn` not as an alternative to `BasicAuth` but as a preferred successor to Form Based Authentication

**Comment by gsmet** [comment]
The commits are semantic, fine by me.

**Comment by cescoffier** [comment]
Hum, this PR update to Vert.x 4.2.6, however, we do not have a constructor run for it yet, and the constructor reported issues with Vert.x main. 

Let's see how it goes, I would need to find slots on between my travels, as I believe we will have a bunch of issues.

**Comment by FroMage** [comment]
If you want to revert this change, I can delay the webauthn support.

**Comment by gsmet** [comment]
Given the fact that Vert.x 4.2.6 seems to be causing problems, I won't backport this to 2.8 for now.

**Comment by cescoffier** [comment]
After discussing with your @Sanne and @DavideD, we are going to revert this PR until we found the problem. Fortunately we will have a vertx 4.2.7 with a fix soon.

**Comment by FroMage** [comment]
Apparently we did a 2.8.1.Final release with Vert.x 4.2.7, so perhaps we can backport this to 2.8 now @gsmet ?

**Comment by gsmet** [comment]
@FroMage FWIW, I preferred waiting for 2.9 as this feature needs proper exposure and micros don't really get that much exposure given people are not expecting new features.

**Pull request merged**
